design/phase-level-progress-tracking.md

```diff
@@ -0,0 +1,143 @@
+# Problem Statement
+
+Kanister actions are triggered by creating ActionSet resource. The status field in ActionSet CR represents the status of Kanister operation. The action completion percentage is set by calculating the weights given on phases ([https://github.com/kanisterio/kanister/blob/master/design/progress-tracking.md](https://github.com/kanisterio/kanister/blob/master/design/progress-tracking.md) ).
+
+This goal of this proposal is
+
+- Have progress tracking per phase
+- Improve action progress tracking mechanism by removing hard coded weights
+
+## High Level Design
+
+### Changes in ActionSet status field
+
+In order to inform progress of each phase execution while performing action, the progress field can be added to status.phase
+
+```
+// ActionStatus is updated as we execute phases.
+type ActionStatus struct {
+        // Name is the action we'll perform. For example: `backup` or `restore`.
+        Name string `json:"name"`
+        // Object refers to the thing we'll perform this action on.
+        Object ObjectReference `json:"object"`
+        // Blueprint with instructions on how to execute this action.
+        Blueprint string `json:"blueprint"`
+        // Phases are sub-actions an are executed sequentially.
+        Phases []Phase `json:"phases,omitempty"`
+        // Artifacts created by this phase.
+        Artifacts map[string]Artifact `json:"artifacts,omitempty"`
+        // DeferPhase is the phase that is executed at the end of an action
+        // irrespective of the status of other phases in the action
+        DeferPhase Phase `json:"deferPhase,omitempty"`
+}
+
+ // Phase is subcomponent of an action.
+ type Phase struct {
+        Name     string                 `json:"name"`
+        State    State                  `json:"state"`
+        Output   map[string]interface{} `json:"output,omitempty"`
++       Progress PhaseProgress          `json:"progress,omitempty"` <-----
+ }
+
++type PhaseProgress struct {
++       ProgressPercent       int64
++       LastTransitionTime    *metav1.Time
++       SizeUploadedB         int64
++       EstimatedTimeSeconds  int64
++       EstimatedUploadSizeB  int64
++}
+```
+
+### Action progress tracking
+
+Action can consists of multiple phases. These all phases can take consume different duration for completion based on the operations it performs. All phases for the action can be given equal weight-age and to calculate action progress completion %, average of all phase completion %ge can be calculated
+
+```
+% completion of action = sum(% Phase completion of all the phases)/ no. of phases
```

Comment:
Are we not including the combined result as part of `ActionStatus`?

Reply:
We have ActionSetStatus.Progress and Phase level Progress, but not ActionStatus since use can only see actionset status. Is that the reason progress of each action is not included?

Reply:
Since ActionSet is supposed to perform single action only, the ActionSet progress is the same as Action progress. So by ActionSet.Progress users can get information about Action progress as well.